docs/design-v2.md

```diff
@@ -0,0 +1,132 @@
+plug-in runs. An instance of `AzDriverNode` is created when the node plug-in starts. The node 
+plug-in periodically updates the heartbeat in the `Status` field.
+
+The controller for `AzDriverNode` runs in the controller plug-in. It is responsible for 
+deleting `AzDriverNode` instances that no longer have corresponding nodes in the cluster.
+
+#### `AzVolume` Resource and Controller
+
+The `AzVolume` custom resource represents the managed disk of a `PersistentVolume`. The 
+controller for `AzVolume` runs in the controller plug-in. It watches for and reconciles 
+changes in the `AzVolume` instances.
+
+An `AzVolume` instance is created by the `CreateVolume` API to the CSI Controller plug-in. The 
+`AzVolume` controller responds to the new instance by creating a managed disk using the 
+parameters in the referenced `StorageClass`. When disk creation is complete, the controller 
+sets the `.Status.State` field to `Created` or suitable error state on failure. It also adds a 
+finalizer if creation was successful. The `CreateVolume` request completes when the `AzVolume` 
+state is updated.
+
+To delete a managed disk, the `DeleteVolume` API in the CSI Controller plug-in deletes the 
+corresponding `AzVolume` instance. The controller responds to deletion by garbage collecting 
+the managed disk. When the managed disk has been deleted, the controller removes the finalizer
+from the `AzVolume` instance. The `DeleteVolume` request completes once the `AzVolume` instance
+has been removed from the object store.
+
+#### `AzVolumeAttachment` Resource and Controller
+
+The `AzVolumeAttachment` custom resource represents the attachment of a managed disk to a 
+specific node. The controller for this custom resource runs in the controller plug-in and 
+watches for changes in the `AzVolumeAttachment` instances.
+
+An `AzVolumeAttachment` instance representing the primary node attachment is created by the 
+`ControllerPublishVolume` API in the CSI Controller plug-in. If an instance for the current 
+node already exists, it is updated to represent the primary node attachment. The 
+`AzVolumeAttachment` controller then attaches the manage disk to the primary node (downgrading 
+to mount replica or detaching any previous primary attachment if one exists). It then creates 
+additional `AzVolumeAttachment` instances representing mount replicas and attaches the shared 
+manage disk to a number of backup nodes as specified by the `maxMountReplicaCount` parameter 
+in the `StorageClass` instance of the `PersistentVolumeClaim` of the managed disk. The 
+controller also sets the `.Status.State` field to `Attached` and a finalizer to each 
+`AzVolumeAttachment` instance as each attachment completes. The `ControllerPublishVolume` 
+request is complete when the state of the primary node attachment is complete.
+
+When the `ControllerUnpublishVolume` API in the CSI Controller plug-in is called, it deletes 
```

Comment:
I am not sure if this could be a valid scenario but consider If the pod needs to be rescheduled to a different node because the current node has become unresponsive. The VolumeAttachment object pointing to that node will be  deleted by the K8s controllers because the node is unresponsive and the deletion of VolumeAttachment will trigger ControllerUnpublishVolume.  

In that case, does it make sense to detach the disk from the mount replica nodes? Wouldn't it be better to keep the replica nodes and make one of them as the primary node and schedule the pod over there?

Reply:
For attach code path, behavior is what Ed describes.

For failover code path:
1. ControllerUnpublishVolume should delete the AzVolumeAttachment and the controller should force release the attachment and remove the finalizer on AzVolumeAttachment.
2. The following ControllerPublishVolume call should ideally come for a node where we already have the attachment. In anycase, we should either create a new AzVolumeAttachment or edit the existing AzVolumeAttachment for the node and mark it as primary, or change the version somehow so that controller tries to reconcile the attachment. This will ensure that, if someone goes and accidently deletes the attachment from controller, our controller would still be able to reconcile.

For volume delete code path:
1. This is when all the back-up attachments should get deleted.

Reply:
We will potentially need attachment rebalancing as well as part of the placement logic to make sure, all volumes have healthy replica counts and replicas are evenly distributed.

Reply:
I update the document so that it no longer unconditionally deletes the `AzVolumeAttachment` instances in `ControllerUnpublishVolume`. However, I have left it unspecified how we determine and handle failover vs. volume delete cases. I suppose the next call for the volume with either be a `ControllerPublishVolume` (failover) or `ControllerDeleteVolume` (delete for dynamic provisioned disks). We could use those as cues to remove the attachment replicas. I guess we just need to fall back to garbage collection for the static provisioning case.